canopen_ros2_control/src/robot_system.cpp

```diff
@@ -0,0 +1,308 @@
+        break;
+      case MotorBase::Profiled_Position:
+      case MotorBase::Cyclic_Synchronous_Position:
+      case MotorBase::Interpolated_Position:
+        motion_controller_driver->set_target(motor_data_[it->first].target_position);
+        break;
+      case MotorBase::Profiled_Velocity:
+      case MotorBase::Cyclic_Synchronous_Velocity:
+        motion_controller_driver->set_target(motor_data_[it->first].target_velocity);
+        break;
+      case MotorBase::Profiled_Torque:
+        motion_controller_driver->set_target(motor_data_[it->first].target_torque);
+        break;
+      default:
+        RCLCPP_INFO(rclcpp::get_logger("robot_system_interface"), "Mode not supported");
+    }
+  }
+
+  return hardware_interface::return_type::OK;
+}
+
+void Cia402RobotSystem::initDeviceContainer()
+{
+  std::string tmp_master_bin = (info_.hardware_parameters["master_bin"] == "\"\"")
+                                 ? ""
+                                 : info_.hardware_parameters["master_bin"];
+
+  device_container_->init(
+    info_.hardware_parameters["can_interface_name"], info_.hardware_parameters["master_config"],
+    info_.hardware_parameters["bus_config"], tmp_master_bin);
+  auto drivers = device_container_->get_registered_drivers();
+  RCLCPP_INFO(
+    rclcpp::get_logger("robot_system_interface"), "Number of registered drivers: '%zu'",
+    device_container_->count_drivers());
+  for (auto it = drivers.begin(); it != drivers.end(); it++)
+  {
+    uint16_t node_id = it->first;
+    std::string driver_type = device_container_->get_driver_type(node_id);
+    RCLCPP_INFO(
+      robot_system_logger, "Found driver type '%s' at node_id '%u'", driver_type.c_str(), node_id);
+
+    // Check if driver at id is of type ros2_canopen::Cia402Driver
+    if (driver_type.compare("ros2_canopen::Cia402Driver") == 0)
+    {
+      cia402_drivers_.emplace_back(
+        std::static_pointer_cast<ros2_canopen::Cia402Driver>(it->second));
+    }
+    else
+    {
+      RCLCPP_ERROR(
+        robot_system_logger, "Driver at node_id '%u' is not of type ros2_canopen::Cia402Driver",
+        node_id);
+    }
+  }
+  RCLCPP_INFO(device_container_->get_logger(), "Initialisation successful.");
+}
+
+void Cia402RobotSystem::spin()
+{
+  executor_->spin();
+  executor_->remove_node(device_container_);
+  RCLCPP_INFO(device_container_->get_logger(), "Stopped spinning Cia402RobotSystem ROS2 executor");
+}
+
+void Cia402RobotSystem::clean()
+{
+  printf("Cancel exectutor...");
+  executor_->cancel();
+  printf("Join spin thread...");
+  spin_thread_->join();
+
+  printf("Reset variables...");
+  device_container_.reset();
+  executor_.reset();
+
+  init_thread_->join();
+  init_thread_.reset();
+
+  executor_.reset();
+  spin_thread_.reset();
+  motor_data_.clear();
+  cia402_drivers_.clear();
+}
+
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(canopen_ros2_control::Cia402RobotSystem, hardware_interface::SystemInterface)
```

Comment:
Most of this file is copied from [here](https://github.com/ros-industrial/ros2_canopen/blob/master/canopen_ros2_control/src/cia402_system.cpp), but there is no license mention nor the any link about the authors...